rust_snuba/src/strategies/python.rs

```diff
@@ -13,131 +17,131 @@ use crate::config::MessageProcessorConfig;
     where
         N: ProcessingStrategy<BytesInsertBatch> + 'static,
     {
         let next_step = Box::new(next_step);
         let python_module = &processor_config.python_module;
         let python_class_name = &processor_config.python_class_name;
-        let code = format!(
-            r#"
-import rapidjson
-from snuba.datasets.processors import DatasetMessageProcessor
-
-from {python_module} import {python_class_name} as Processor
-
-processor = Processor.from_kwargs()
-
-from snuba.consumers.types import KafkaMessageMetadata
-from snuba.processor import InsertBatch
-from datetime import datetime
-from snuba.consumers.consumer import json_row_encoder
-
-
-def _wrapped(message, offset, partition, timestamp):
-    rv = processor.process_message(
-        message=rapidjson.loads(bytearray(message)),
-        metadata=KafkaMessageMetadata(
-            offset=offset,
-            partition=partition,
-            timestamp=timestamp
-        )
-    )
-
-    if rv is None:
-        return rv
-
-    assert isinstance(rv, InsertBatch), "this consumer does not support replacements"
-
-    return [json_row_encoder.encode(row) for row in rv.rows]
-"#
-        );
-
-        let py_process_message = Python::with_gil(|py| -> PyResult<Py<PyAny>> {
-            let fun: Py<PyAny> = PyModule::from_code(py, &code, "", "")?
-                .getattr("_wrapped")?
-                .into();
-            Ok(fun)
-        })?;
 
         Ok(PythonTransformStep {
             next_step,
-            py_process_message,
+            handles: VecDeque::new(),
             message_carried_over: None,
+            processing_pool: procspawn::Pool::builder(processes)
+                .env("RUST_SNUBA_PROCESSOR_MODULE", python_module)
+                .env("RUST_SNUBA_PROCESSOR_CLASSNAME", python_class_name)
+                .build().expect("failed to build procspawn pool"),
         })
     }
-}
 
-impl ProcessingStrategy<KafkaPayload> for PythonTransformStep {
-    fn poll(&mut self) -> Option<CommitRequest> {
-        if let Some(message) = self.message_carried_over.take() {
-            if let Err(MessageRejected {
-                message: transformed_message,
-            }) = self.next_step.submit(message)
-            {
-                self.message_carried_over = Some(transformed_message);
+    fn check_for_results(&mut self) {
+        // procspawn has no join() timeout that does not consume the handle on timeout.
+        //
+        // Additionally we have observed, at least on MacOS, that procspawn's active_count() only
+        // decreases when the handle is consumed. Therefore our count of actually saturated
+        // processes is `self.processing_pool.active_count() - self.handles.len()`.
+        //
+        // If no process is saturated (i.e. above equation is <= 0), we can conclude that all tasks
+        // are done and all handles can be joined and consumed without waiting.
```

Comment:
@mitsuhiko does this reasoning make sense to you? I am not sure what the intended behavior of `active_count` is

Reply:
I think that logic makes sense but honestly this probably needs changing in procspawn.